### PR TITLE
[function table] Make sure FunctionsToRun are executed properly on all workers

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
This PR fix the issue that sometimes FunctionsToRun is not executed. We isolated the Functions/Actors in function table, but not the RunctionsToRun. So when doing importing, sometimes, some functions will be missed.
This PR fixed this.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/worker.py [hunk 1]
```diff
@@ -390,31 +390,24 @@ def run_function_on_all_workers(self, function):
 
             function_to_run_id = hashlib.shake_128(pickled_function).digest(
                 ray_constants.ID_SIZE)
-            key = b"FunctionsToRun:" + function_to_run_id
+            key = b"FunctionsToRun:" + self.current_job_id.hex().encode(
```
Comment (ericl) [inline python/ray/worker.py:393]: Can this be pulled out into a utility function?

File: python/ray/worker.py [hunk 2]
```diff
@@ -390,31 +390,24 @@ def run_function_on_all_workers(self, function):
...
+            ) + b":" + function_to_run_id
             # First run the function on the driver.
             # We always run the task locally.
             function({"worker": self})
-            # Check if the function has already been put into redis.
```
Comment (ericl) [inline python/ray/worker.py:397]: Do we no longer need to check the duplicate?

**Review by ericl — COMMENTED** [review]
What was the symptom of sometimes missing functions? If I understand, it was missing the job isolation prefix, so would we never import any of them?

File: python/ray/worker.py [hunk 3]
```diff
@@ -390,31 +390,24 @@ def run_function_on_all_workers(self, function):
...
             # Run the function on all workers.
-            self.gcs_client.internal_kv_put(
-                key,
-                pickle.dumps({
-                    "job_id": self.current_job_id.binary(),
-                    "function_id": function_to_run_id,
-                    "function": pickled_function,
-                }), True, ray_constants.KV_NAMESPACE_FUNCTION_TABLE)
-            self.function_actor_manager.export_key(key)
+            if self.gcs_client.internal_kv_put(
```
Comment (iycheng) [inline python/ray/worker.py:403]: @ericl  we still check here.

File: python/ray/tests/test_basic_5.py [hunk 4]
```diff
@@ -118,5 +119,41 @@ def test_internal_kv(ray_start_regular):
...
+def init_func(worker_info):
+    a = ray.get_actor("recorder", namespace="n")
+    a.record.remote(worker_info['worker'].worker_id)
+
+ray.worker.global_worker.run_function_on_all_workers(init_func)
+ray.init(address='auto')
+@ray.remote
+def ready():
+    a = ray.get_actor("recorder", namespace="n")
+    assert ray.worker.global_worker.worker_id in ray.get(a.record.remote())
```
Comment (iycheng) [inline python/ray/tests/test_basic_5.py:149]: > What was the symptom of sometimes missing functions? If I understand, it was missing the job isolation prefix, so would we never import any of them?

@ericl  yes, we'll never import any of them. So, this assertion will fail without this PR.

**Comment by iycheng** [comment]
It seems to break runtime env. I'll take a look

File: python/ray/tests/test_basic_5.py (hunk 4 above)
Comment (ericl) [reply python/ray/tests/test_basic_5.py:149]: I see, is run_function_on_all_workers even an API that we support though? What uses this / would break?

Comment (ericl) [reply python/ray/tests/test_basic_5.py:149]: (Is this something that should block the release?)

Comment (iycheng) [reply python/ray/tests/test_basic_5.py:149]: I think so. It's not a public API, but we are using them here

https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/worker.py?L1578-1581

It'll be used to change the working dir, so for runtime env, I think it'll fail in some cases. I can try to get a script failure.

**Review by ericl — APPROVED** [review]

**Pull request merged**